sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java

```diff
@@ -1572,6 +1588,63 @@ public int compare(Method o1, Method o2) {
     return builder.build();
   }
 
+  /**
+   * Attempt to parse an input string into an instance of `type` using an {@link ObjectMapper}.
+   *
+   * <p>If the getter method is annotated with {@link
+   * com.fasterxml.jackson.databind.annotation.JsonDeserialize} the specified deserializer will be
+   * used, otherwise the default ObjectMapper deserialization strategy.
+   *
+   * <p>Parsing is attempted twice, once with the raw string value. If that attempt fails, another
+   * attempt is made by wrapping the value in quotes so that it is interpreted as a JSON string.
+   */
+  private static Object tryParseObject(String value, JavaType type, Method method)
+      throws IOException {
+    AnnotationCollector ac = AnnotationCollector.emptyCollector();
+    for (Annotation ann : method.getAnnotations()) {
+      ac = ac.addOrOverride(ann);
+    }
+
+    AnnotatedMethod annotatedMethod =
+        new AnnotatedMethod(
+            new TypeResolutionContext.Empty(MAPPER.getTypeFactory()),
+            method,
+            ac.asAnnotationMap(),
+            null);
+
+    BeanPropertyDefinition propDef =
+        SimpleBeanPropertyDefinition.construct(MAPPER.getDeserializationConfig(), annotatedMethod);
+
+    BeanProperty prop =
+        new MethodProperty(
+            propDef,
+            type,
+            MAPPER.getDeserializationConfig().findTypeDeserializer(type),
+            ac.asAnnotations(),
+            annotatedMethod);
+
+    JsonNode tree;
+    try {
+      tree = MAPPER.readTree(value);
+    } catch (JsonParseException e) {
+      // try again, quoting the input string if it wasn't already
+      if (!(value.startsWith("\"") && value.endsWith("\""))) {
+        try {
+          tree = MAPPER.readTree("\"" + value + "\"");
+        } catch (JsonParseException inner) {
+          // rethrow the original exception rather the one thrown from the fallback attempt
+          throw e;
+        }
+      } else {
+        throw e;
+      }
+    }
+
+    JsonParser parser = new TreeTraversingParser(tree, MAPPER);
+    parser.nextToken();
+    return DESERIALIZATION_CONTEXT.readPropertyValue(parser, prop, type);
```

Comment:
As a point of conversation, I'm curious what people think of using the built-in jackson `@JsonDeserialize` vs making our own attribute?  The downside I see of the built-in attribute is that other things may parse pipeline options (dataflow for example writes it out as json), and this will affect that deserialization as well.

We could have a `@ParseFunction` attribute or similar to map String -> T instead possibly, which would only be used from PipelineOptionsFactory.

Curious what people's thoughts are.

Reply:
This sounds like a useful improvement but we'll want `@JsonSerialize` as well since we need the value to round trip to JSON and back (unless the getter is marked with `@JsonIgnore` in that case we don't need the `@JsonSerialize`). 

But be aware that you'll want to ensure that all instances used via the "as" transform use the same annotation since multiple PipelineOptions can define the same property effectively and we would want to limit how inconsistent conversions can appear so this will limit the feature a little.
For example with `@JsonIgnore` https://github.com/apache/beam/blob/2ab002486345f5d97f11ccfbeb86541796c080e0/sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java#L941 
https://github.com/apache/beam/blob/2ab002486345f5d97f11ccfbeb86541796c080e0/sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java#L1054

Reply:
I'll double check, but I think it should already be supported.  The places I've seen pass the options into the mapper directly, which will use the annotations.  We only need this workaround here since the PipelineOptions class isn't directly deserialized.

Reply:
> But be aware that you'll want to ensure that all instances used via the "as" transform use the same annotation

ack, I had saw this but didn't implement it yet pending agreement on using the jackson annotation.  I'll add it in.